### PR TITLE
CUMULUS-2631: Add Bearer token support to s3credentials endpoint

### PR DESCRIPTION
**Pull request by jennyhliu**
**Summary:** Summary of changes

Addresses [CUMULUS-2632: Add Bearer token support to s3credentials endpoint](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2631)

## Changes

* Add Bearer token support to s3credentials endpoint
* document PR https://github.com/nasa/cumulus-distribution-api/pull/8

## Ad-hoc testing

1. Get a EDL token, see https://wiki.earthdata.nasa.gov/display/EL/API+Documentation#APIDocumentation-oauth-tokens-user
    echo -n "username:passwor”d | base64
    curl --request POST "https://uat.urs.earthdata.nasa.gov/api/users/token" -H "Authorization: Basic base64-text-from-previous-command"
 2. curl https://tea-api-endpoint.execute-api.us-east-1.amazonaws.com:portconfigured-suchas-7000/DEV/s3credentials --header 'Authorization: Bearer edl-token'

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Ad-hoc testing - Deploy changes and test manually
- [x] Integration tests

### CODE REVIEW

File: example/spec/parallel/s3Access/teaS3CredentialsSpec.js [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    beforeAll(async () => {
+      accessTokensModel = new AccessToken();
+      username = process.env.EARTHDATA_USERNAME;
+      accessTokenResponse = await getEarthdataAccessToken({
+        redirectUri: process.env.DISTRIBUTION_REDIRECT_ENDPOINT,
+        requestOrigin: process.env.DISTRIBUTION_ENDPOINT,
+        storeAccessToken: false,
+      }).catch((error) => {
+        console.log(error);
+        throw error;
```
Comment (botanical) [inline example/spec/parallel/s3Access/teaS3CredentialsSpec.js:51]: Will this actually throw and stop running everything else if there is an error? I think the pattern we having been doing as a workaround to this jasmine issue is 
```
if (beforeAllFailed) fail('beforeAll() failed');
``` 

in the `it` block

File: packages/integration-tests/api/distribution.js [hunk 2]
```diff
@@ -106,30 +106,22 @@ async function invokeTEADistributionLambda(
 }
 
 /**
- * Invoke s3-credentials-endpoint lambda directly to get s3 credentials. This
- * is used in integration testing so that we use the lambda's IAM
- * role/permissions when accessing resources.
+ * Invoke s3-credentials-endpoint lambda directly to get s3 credentials.=
```
Comment (botanical) [inline packages/integration-tests/api/distribution.js:109]: Was this change an accident?

File: packages/integration-tests/api/distribution.js [hunk 3]
```diff
@@ -106,30 +106,22 @@ async function invokeTEADistributionLambda(
...
  *
- * @param {string} path
- *   path to file requested.  This is just "/bucket/keytofile"
- * @param {string} accessToken
- *   Access token from OAuth provider or nothing.
- * @returns {string}
- *   signed s3 URL for the requested file.
+ * @param {string} path request path
+ * @param {Object} headers requst header
```
Comment (botanical) [inline packages/integration-tests/api/distribution.js:112]:
```suggestion
 * @param {Object} headers request header
```

File: packages/s3-credentials-endpoint/index.js [hunk 4]
```diff
@@ -90,13 +70,13 @@ function isPublicRequest(path) {
 async function handleRedirectRequest(req, res) {
   const {
     accessTokenModel,
-    authClient,
+    oauthClient,
     distributionUrl,
-  } = getConfigurations();
+  } = await getConfigurations();
```
Comment (botanical) [inline packages/s3-credentials-endpoint/index.js:75]: 👍

File: packages/s3-credentials-endpoint/index.js [hunk 5]
```diff
@@ -170,30 +150,36 @@ async function ensureAuthorizedOrRedirect(req, res, next) {
...
-    if (error instanceof RecordDoesNotExist) {
-      return res.redirect(307, redirectURLForAuthorizationCode);
+    if (!(error instanceof RecordDoesNotExist)) {
+      throw error;
     }
-    throw error;
   }
 
-  if (isAccessTokenExpired(accessTokenRecord)) {
+  if (!accessToken || !accessTokenRecord || isAccessTokenExpired(accessTokenRecord)) {
```
Comment (botanical) [inline packages/s3-credentials-endpoint/index.js:182]: `if (!accessToken ||...)`
is handled on line above https://github.com/nasa/cumulus/pull/3078/files#diff-3bb3b5cbe8f7561528077af0e7c41cac9db4d4b8728f061d53dfec432fab8c61R171 right?

File: packages/s3-credentials-endpoint/tests/test-index.js [hunk 6]
```diff
@@ -216,3 +238,51 @@ test.serial('An s3credential request with DISABLE_S3_CREDENTIALS set to true res
   });
 });
 /* eslint-enable lodash/prefer-noop */
+
+test.serial('An s3credential request with a valid bearer token invokes NGAPs request for credentials with username from token ', async (t) => {
```
Comment (botanical) [inline packages/s3-credentials-endpoint/tests/test-index.js:242]:
```suggestion
test.serial('An s3credential request with a valid bearer token invokes NGAPs request for credentials with username from token', async (t) => {
```

File: packages/s3-credentials-endpoint/tests/test-index.js [hunk 7]
```diff
@@ -216,3 +238,51 @@ test.serial('An s3credential request with DISABLE_S3_CREDENTIALS set to true res
...
+
+  t.true(spy.called);
+  t.deepEqual(spy.args[0][0], {
+    FunctionName,
+    Payload,
+  });
+  stub.restore();
+});
+
+test('An s3credential request with a invalid bearer token redirects to Oauth2 provider', async (t) => {
```
Comment (botanical) [inline packages/s3-credentials-endpoint/tests/test-index.js:277]:
```suggestion
test('An s3credential request with an invalid bearer token redirects to OAuth2 provider', async (t) => {
```

File: tf-modules/distribution/main.tf [hunk 8]
```diff
@@ -136,7 +136,7 @@ resource "aws_lambda_function" "s3_credentials" {
   role             = aws_iam_role.s3_credentials_lambda[0].arn
   runtime          = "nodejs14.x"
   timeout          = 50
-  memory_size      = 320
+  memory_size      = 512
```
Comment (botanical) [inline tf-modules/distribution/main.tf:139]: 👍

File: tf-modules/distribution/outputs.tf [hunk 9]
```diff
@@ -1,5 +1,5 @@
 output "s3_credentials_redirect_uri" {
-  value = var.deploy_s3_credentials_endpoint ? "${var.tea_internal_api_endpoint}redirect" : null
```
Comment (botanical) [inline tf-modules/distribution/outputs.tf:2]: Why did this use internal api endpoint before?

Comment (jennyhliu) [reply tf-modules/distribution/outputs.tf:2]: Not sure.  But it doesn't work if I want to add port to the tea distribution url.
`tea_internal_api_endpoint` is the api gateway endpoint. 
`tea_external_api_endpoint` is the tea distribution endpoint which can be set with terraform variable `tea_distribution_url`.
e.g.
tea_internal_api_endpoint = https://tea-endpoint.execute-api.us-east-1.amazonaws.com/DEV/
tea_external_api_endpoint = https://tea-endpoint.execute-api.us-east-1.amazonaws.com:7000/DEV/ or cloudfront url

File: packages/s3-credentials-endpoint/index.js (hunk 5 above)
Comment (jennyhliu) [reply packages/s3-credentials-endpoint/index.js:182]: good catch!  Updated

File: packages/integration-tests/api/distribution.js (hunk 2 above)
Comment (jennyhliu) [reply packages/integration-tests/api/distribution.js]: No.  I feel the sentence `This is used in..` is redundant

**Review by botanical — APPROVED** [review]
Nice integration tests! Thanks for addressing the feedback

**Pull request merged**